File: api/xSOLACEv1/circulatingSupply/get.py
from api.utils import *

def getCirculatingSupply():
    w3 = Web3(Web3.HTTPProvider("https://eth-mainnet.alchemyapi.io/v2/{}".format(alchemy_key)))
    xsolace = w3.eth.contract(address=xSOLACEv1_ADDRESS, abi=erc20Json)
    blocknum = w3.eth.blockNumber
    supply = xsolace.functions.totalSupply().call(block_identifier=blocknum)
    skip_addresses = json.loads(s3_get("xSOLACEv1/circulatingSupply/skip_addresses.json", cache=True))
    for addr in skip_addresses:
        supply -= xsolace.functions.balanceOf(addr).call(block_identifier=blocknum)
    supplyNormalized = supply / ONE_ETHER
    return supplyNormalized

def handler(event, context):
    try:
        supply = getCirculatingSupply()
        return {
            "statusCode": 200,
            "body": supply,
            "headers": headers
        }
    except InputException as e:
        return handle_error(event, e, 400)
    except Exception as e:
        return handle_error(event, e, 500)

File: api/xSOLACEv1/circulatingSupply/get.js
const { getProvider, s3GetObjectPromise, snsPublishError } = require("./../../utils/utils")
const ethers = require('ethers')
const BN = ethers.BigNumber
const formatUnits = ethers.utils.formatUnits

// Define headers
const headers = {
  "Content-Type": "application/json",
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Headers": "Content-Type",
  "Access-Control-Allow-Methods": "OPTIONS,GET,POST,PUT,DELETE"
}

const XSOLACE_V1_ADDRESS = "0x501AcE5aC3Af20F49D53242B6D208f3B91cfc411"
const ERC20ABI = [{"inputs":[],"name":"totalSupply","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"account","type":"address"}],"name":"balanceOf","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"}]

async function getCirculatingSupply() {
  var [provider] = await Promise.all([
    getProvider(1)
  ])
  var xsolacev1 = new ethers.Contract(XSOLACE_V1_ADDRESS, ERC20ABI, provider)
  var supply = await xsolacev1.totalSupply()
  return supply
}

async function handle(event) {
  var supply = await getCirculatingSupply()
  return formatUnits(supply, 18)
}

// Lambda handler
exports.handler = async function(event) {
  try {
    var res = await handle(event)
    return {
      statusCode: 200,
      headers: headers,
      body: res
    }
  } catch (e) {
    switch(e.name) {
      case "InputError":
        return {
          statusCode: 400,
          headers: headers,
          body: e.stack
        }
        break
      default:
        await snsPublishError(event, e)
        return {
          statusCode: 500,
          headers: headers,
          body: "internal server error"
        }
    }
  }
}
